### PR TITLE
feat(Core/Conf/BG): Allow battlegrounds to be configurable (#20320)

### PR DESCRIPTION
**Pull request by valsan-azerty-boi**
## Changes Proposed:
This PR proposes changes to:
-  [x] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/20320
image

#### Warsong
- Configurable number of flags to win => OK
#### Arathi
- Configurable captures points to win => OK
#### Alterac
- Configurable number of reinforcements => OK
- Configurable boss reputation on boss death => OK
#### Eye of the Storm
- Configurable captures points to win => OK
**Note: No WorldState exists for the max score in this battlefield (XXXX/1600), and  I found nothing elsewhere for wotlk
so if you set the maxscore to 2000 for example 
it will display scores likes 1700/1600 or 2000/1600 instead of 1700/2000 or 2000/2000, seems client related,
the problem exists only on Eye battleground (no problem in Arathi)**
#### Strand of the Ancients
- No modification => Nothing interesting to put in configuration
#### Isle of Conquest
- No modification => Weird side effects with modified reinforcements, 
and I don't want to make too much modifications to the existing code,
so I let it unmodified and non configurable for now

## SOURCE:
The changes have been validated through:
- [x] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:
This PR has been:
- [x] Tested in-game by the author.
- [x] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:
- [x] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1. Tested with original & modified Warsong max flags from configuration
warsong_custom_Flags warsong_custom_Flags2 image

2. Tested with original & modified Arathi max capture points from configuration
arathi_custom_CapturePoints arathi_custom_CapturePoints2 image image

3. Tested with original & modified Alterac reinforcements from configuration
alterac_custom_Reinforcements image

4. Tested with original & modified Alterac boss death rep points from configuration
alterac_custom_RepOnBossDeath2 alterac_custom_RepOnBossDeath image

5. Tested with original & modified Eye of the Storm capture points from configuration
ey_custom_CapturesPoints2 image
**See notes above for Eye of the Storm, works well but not UI compliant**

## Known Issues and TODO List:
- [ ] Eventually make Isle of Conquest configurable (need more code investigations)

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by valsan-azerty-boi** [comment]
Don't hesitate to make a lot of comments if needed, I don"t do C++ very often

File: src/server/apps/worldserver/worldserver.conf.dist [hunk 1]
```diff
@@ -3708,6 +3708,50 @@ Battleground.SpeedBuffRespawn = 150
 
 Battleground.Override.LowLevels.MinPlayers = 0
 
+#
+#    Battleground.Warsong.Flags
+#        Description: Set the number of flags required for a team to win in Warsong battleground
+#        Default:     3 (Blizzlike)
```
Comment (Exitare) [inline src/server/apps/worldserver/worldserver.conf.dist:3714]: Default or Wotlk might be a better name?

File: src/server/game/Battlegrounds/Zones/BattlegroundAB.cpp [hunk 2]
```diff
@@ -478,6 +478,11 @@ void BattlegroundAB::Init()
     _capturePointInfo[BG_AB_NODE_BLACKSMITH]._iconCapture = BG_AB_OP_BLACKSMITH_STATE_ALIENCE;
     _capturePointInfo[BG_AB_NODE_LUMBER_MILL]._iconCapture = BG_AB_OP_LUMBERMILL_STATE_ALIENCE;
     _capturePointInfo[BG_AB_NODE_GOLD_MINE]._iconCapture = BG_AB_OP_GOLDMINE_STATE_ALIENCE;
+
+    int bgArathiCapturePointsConfig = sWorld->getIntConfig(CONFIG_BATTLEGROUND_ARATHI_CAPTUREPOINTS);
```
Comment (Exitare) [inline src/server/game/Battlegrounds/Zones/BattlegroundAB.cpp:482]: This can be a uint

File: src/server/game/Battlegrounds/Zones/BattlegroundEY.cpp [hunk 3]
```diff
@@ -361,6 +361,11 @@ void BattlegroundEY::Init()
     _droppedFlagGUID.Clear();
     _flagState = BG_EY_FLAG_STATE_ON_BASE;
     _flagCapturedObject = 0;
+
+    int bgEyCapturePointsConfig = sWorld->getIntConfig(CONFIG_BATTLEGROUND_EYEOFTHESTORM_CAPTUREPOINTS);
```
Comment (Exitare) [inline src/server/game/Battlegrounds/Zones/BattlegroundEY.cpp:365]: uint

File: src/server/game/Battlegrounds/Zones/BattlegroundWS.cpp [hunk 4]
```diff
@@ -497,6 +497,11 @@ void BattlegroundWS::Init()
         _honorWinKills = 1;
         _honorEndKills = 2;
     }
+
+    int bgWarsongFlagsConfig = sWorld->getIntConfig(CONFIG_BATTLEGROUND_WARSONG_FLAGS);
```
Comment (Exitare) [inline src/server/game/Battlegrounds/Zones/BattlegroundWS.cpp:501]: uint

File: src/server/game/World/World.cpp [hunk 5]
```diff
@@ -1172,6 +1172,12 @@ void World::LoadConfigSettings(bool reload)
         _int_configs[CONFIG_BATTLEGROUND_SPEED_BUFF_RESPAWN] = 150;
     }
 
+    _int_configs[CONFIG_BATTLEGROUND_WARSONG_FLAGS]                        = sConfigMgr->GetOption<uint32>("Battleground.Warsong.Flags", 3);
+    _int_configs[CONFIG_BATTLEGROUND_ARATHI_CAPTUREPOINTS]                 = sConfigMgr->GetOption<uint32>("Battleground.Arathi.CapturePoints", 1600);
+    _int_configs[CONFIG_BATTLEGROUND_ALTERAC_REINFORCEMENTS]               = sConfigMgr->GetOption<int32>("Battleground.Alterac.Reinforcements", 600);
+    _int_configs[CONFIG_BATTLEGROUND_ALTERAC_REP_ONBOSSDEATH]              = sConfigMgr->GetOption<int32>("Battleground.Alterac.ReputationOnBossDeath", 350);
```
Comment (Exitare) [inline src/server/game/World/World.cpp:1178]: uint

File: src/server/apps/worldserver/worldserver.conf.dist (hunk 1 above)
Comment (valsan-azerty-boi) [reply src/server/apps/worldserver/worldserver.conf.dist:3714]: What do you prefere ?

Comment (valsan-azerty-boi) [reply src/server/apps/worldserver/worldserver.conf.dist:3714]: I have found a lot of occurence of (Blizzlike) for options in configuration so I think it's good

Comment (valsan-azerty-boi) [reply src/server/apps/worldserver/worldserver.conf.dist:3714]: Ok I understand !
But for me in the case of AzerothCore: Blizzlike = WotLK
but it's my interpretation lol :p

File: src/server/game/Battlegrounds/Zones/BattlegroundAB.cpp (hunk 2 above)
Comment (valsan-azerty-boi) [reply src/server/game/Battlegrounds/Zones/BattlegroundAB.cpp]: fixed in commit 594fb62c56be6468a2d0175082327a562f7a9f9b

File: src/server/game/Battlegrounds/Zones/BattlegroundEY.cpp (hunk 3 above)
Comment (valsan-azerty-boi) [reply src/server/game/Battlegrounds/Zones/BattlegroundEY.cpp]: fixed in commit 594fb62c56be6468a2d0175082327a562f7a9f9b

File: src/server/game/Battlegrounds/Zones/BattlegroundWS.cpp (hunk 4 above)
Comment (valsan-azerty-boi) [reply src/server/game/Battlegrounds/Zones/BattlegroundWS.cpp]: fixed in commit 594fb62c56be6468a2d0175082327a562f7a9f9b

File: src/server/game/World/World.cpp (hunk 5 above)
Comment (valsan-azerty-boi) [reply src/server/game/World/World.cpp]: fixed in commit 594fb62c56be6468a2d0175082327a562f7a9f9b

File: src/server/apps/worldserver/worldserver.conf.dist (hunk 1 above)
Comment (Exitare) [reply src/server/apps/worldserver/worldserver.conf.dist:3714]: I understand. But others might not be so versed with that :)

Comment (valsan-azerty-boi) [reply src/server/apps/worldserver/worldserver.conf.dist:3714]: So maybe WotLK will be the most accurate?

Comment (valsan-azerty-boi) [reply src/server/apps/worldserver/worldserver.conf.dist:3714]: As you prefere

Comment (valsan-azerty-boi) [reply src/server/apps/worldserver/worldserver.conf.dist:3714]: or something like that

#Default:     3
#                  1 (Minimum)

(no details after the default: 3)

File: src/server/game/Battlegrounds/Battleground.h [hunk 6]
```diff
@@ -176,9 +176,9 @@ enum BattlegroundTimeIntervals
 
 enum BattlegroundStartTimeIntervals
 {
-    BG_START_DELAY_2M               = 120000,               // ms (2 minutes)
-    BG_START_DELAY_1M               = 60000,                // ms (1 minute)
-    BG_START_DELAY_30S              = 30000,                // ms (30 seconds)
+    BG_START_DELAY_2M               = 5000,               // ms (2 minutes)
+    BG_START_DELAY_1M               = 5000,                // ms (1 minute)
+    BG_START_DELAY_30S              = 5000,                // ms (30 seconds)
     BG_START_DELAY_15S              = 15000,                // ms (15 seconds) Used only in arena
```
Comment (Exitare) [inline src/server/game/Battlegrounds/Battleground.h]: Why did you change it to 5s each?

Comment (valsan-azerty-boi) [reply src/server/game/Battlegrounds/Battleground.h]: no I revert it, was a commit mistake lol (was for test purpose, to make battleground starts faster because was tired of 2min waiting each time lol)

**Comment by Exitare** [comment]
I changed my config for Eye from 1600 to 100. 
i won after I earned 100 points but the visual was still 1600.

**Comment by valsan-azerty-boi** [comment]
Yes, no WorldState in WotLK for the max score XXXX/1600 for Eye of the Storm, can't find anything for that (but no problem for Arathi), so the team score frame stay the same at the top of the screen in Eye of the Storm

**Comment by Exitare** [comment]
Isnt this snippet sending an update to the client? Or is the client just not reacting to it?
`UpdateWorldState(teamId == TEAM_ALLIANCE ? EY_ALLIANCE_RESOURCES : EY_HORDE_RESOURCES, std::min<uint32>(m_TeamScores[teamId], _configurableMaxTeamScore));`

**Comment by valsan-azerty-boi** [comment]
> Isnt this snippet sending an update to the client? Or is the client just not reacting to it? `UpdateWorldState(teamId == TEAM_ALLIANCE ? EY_ALLIANCE_RESOURCES : EY_HORDE_RESOURCES, std::min<uint32>(m_TeamScores[teamId], _configurableMaxTeamScore));

Can't confirm :/

**Comment by valsan-azerty-boi** [comment]
Sorry was sleeping, I try to fix this thing, if I can't I just revert this and keep just conf for:
- Warsong
- Arathi
- Alterac

**Comment by valsan-azerty-boi** [comment]
> Hmm, I would not remove the config for EYE. After all it's a config option that deviates from Blizzlike behavior. We can't fiddle with the client so a note can be added, as you suggested.

I think no one will read the note I added previously in the conf but as you want

**Comment by FrancescoBorzi** [comment]
thank you very much @valsan-azerty-boi for your contribution

**Review by FrancescoBorzi — APPROVED** [review]

**Comment by valsan-azerty-boi** [comment]
> thank you very much @valsan-azerty-boi for your contribution

With pleasure, don't hesitate to ping me if any problems are encountered about this feature
